plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/refactoring/introduce/introduceConstant/KotlinIntroduceConstantHandler.kt

```diff
@@ -0,0 +1,173 @@
+                    }
+                }
+            }
+        }
+    }
+
+    private fun validateElements(elements: List<PsiElement>): String? {
+        val errorMessage = validateExpressionElements(elements)
+        return when {
+            errorMessage != null -> errorMessage
+            elements.any { it.isNotConst() } -> KotlinBundle.message("error.text.can.t.introduce.constant.for.this.expression.because.not.constant")
+            else -> null
+        }
+    }
+
+    private fun PsiElement.isNotConst() = !isConst()
+
+    private fun PsiElement.isConst(): Boolean {
+        if (this is KtBinaryExpression) {
+            var isConst = true
+            object : KotlinRecursiveElementVisitor() {
+                override fun visitKtElement(element: KtElement) {
+                    super.visitKtElement(element)
+                    if (element !is KtOperationReferenceExpression && element !is KtBinaryExpression) {
+                        isConst = isConst && element.isConst()
+                    }
+                }
+            }.visitBinaryExpression(this)
+
+            return isConst
+        } else {
+            return this is KtConstantExpression
+                    || (this is KtStringTemplateExpression && !this.hasInterpolation())
+                    || this is KtLiteralStringTemplateEntry
+        }
+    }
+
+    private fun getNameSuggestions(property: KtProperty): List<String> {
+        val initializerValue = (property.initializer as? KtStringTemplateExpression)?.plainContent
+        val identifierValue = property.identifyingElement?.text
+
+        return listOfNotNull(initializerValue, identifierValue).map { NameUtil.capitalizeAndUnderscore(it) }
+
+    }
+
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
+        if (file !is KtFile) return
+        selectElements(editor, file) { elements, targets -> doInvoke(project, editor, file, elements, targets) }
+    }
+
+    fun selectElements(
+        editor: Editor,
+        file: KtFile,
+        continuation: (elements: List<PsiElement>, targets: PsiElement) -> Unit
+    ) {
+
+        selectElementsWithTargetSibling(
+            INTRODUCE_CONSTANT,
+            editor,
+            file,
+            KotlinBundle.message("title.select.target.code.block"),
+            listOf(CodeInsightUtils.ElementKind.EXPRESSION),
+            ::validateElements,
+            { _, sibling ->
+                sibling.getExtractionContainers(strict = true, includeAll = true)
+                    .filter { (it is KtFile && !it.isScript()) }
+            },
+            continuation
+        )
+    }
+
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        throw AssertionError("$INTRODUCE_CONSTANT can only be invoked from editor")
+    }
+}
+
+val INTRODUCE_CONSTANT: String = KotlinBundle.message("introduce.constant")
```

Comment:
pls be aware of `@Nls`:
```
    val INTRODUCE_CONSTANT
        @Nls
        get() = KotlinBundle.message("introduce.constant")
```

Reply:
Not too sure I completly understood the doc of it, but i added it.